wellcomeml/ml/bert_semantic_equivalence.py

```diff
@@ -0,0 +1,176 @@
+            gen_train,
+            ({"input_ids": tf.int32,
+              "attention_mask": tf.int32,
+              "token_type_ids": tf.int32}, tf.int64),
+            (
+                {
+                    "input_ids": tf.TensorShape([None]),
+                    "attention_mask": tf.TensorShape([None]),
+                    "token_type_ids": tf.TensorShape([None]),
+                },
+                tf.TensorShape([]),
+            ),
+        )
+
+        return dataset
+
+    def fit(self, X, y, random_state=None, epochs=3):
+        """
+        Fits a sentence similarity model
+
+        Args:
+            X: List of 2-uples: [('Sentence 1', 'Sentence 2'), ....]
+            y: list of 0 (not similar) and 1s (similar)
+            random_state: a random state for the train_test_split
+            epochs: number_of epochs
+
+        Returns:
+
+        """
+        # Initialises/downloads model
+        self._initialise_models()
+
+        # Train/val split
+        X_train, X_valid, y_train, y_valid = train_test_split(
+            X, y, test_size=self.test_size, random_state=random_state
+        )
+
+        # Generates tensorflow dataset
+        self.train_dataset = self._prep_dataset_generator(X_train, y_train)
+        self.valid_dataset = self._prep_dataset_generator(X_valid, y_valid)
+
+        # Generates mini-batches and stores in a class variable
+        self.train_dataset = self.train_dataset.shuffle(self.max_length).\
+            batch(self.batch_size).repeat(-1)
+        self.valid_dataset = self.valid_dataset.batch(self.eval_batch_size)
+
+        opt = tf.keras.optimizers.Adam(learning_rate=self.learning_rate,
+                                       epsilon=1e-08)
+
+        loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+
+        # Train and evaluate using tf.keras.Model.fit()
+
+        metric = tf.keras.metrics.SparseCategoricalAccuracy("accuracy")
+
+        self.model.compile(optimizer=opt, loss=loss, metrics=[metric])
+
+        self.train_steps = len(X_train) // self.batch_size
+        self.valid_steps = len(X_valid) // self.eval_batch_size
+
+        self.fit_epoch(epochs=epochs)
+
+        return self
+
+    def fit_epoch(self, epochs=3):
+        """Fits a certain number of epochs"""
+        if not self.train_dataset:
+            raise RuntimeError("You have to first fit a model using .fit")
+
+        self.model.fit(
+            self.train_dataset,
+            epochs=epochs,
+            steps_per_epoch=self.train_steps,
+            validation_data=self.valid_dataset,
+            validation_steps=self.valid_steps,
+        )
+
+        return self
+
+    def score(self, X):
+        X_tokenized = self.tokenizer.batch_encode_plus(
+            X, max_length=self.max_length, add_special_tokens=True,
+            pad_to_max_length=True,
+            return_tensors="tf"
+        )
+        predictions = tf.convert_to_tensor(self.model.predict(X_tokenized))
```

Comment:
predict should probably operate on text as well.

Reply:
the class predict does operate on text. So for `self.predict` you pass a list o 2-uples of texts and it will output whether the text is related (0) or not (1). The tensorflow model (which the end-user will probably not have to use) receives tokenised texts.

I added a docstring in https://github.com/wellcometrust/WellcomeML/pull/43/commits/eb5b9d25d6962ae44ecb52e51f6114a089cca10c for clarification.